POLIO-1645 POLIO-1646  POLIO-1647  incident reports improvements

beygorghor opened the pull request:
Explain what problem this PR is resolving

Related JIRA tickets : POLIO-1645 POLIO-1646  POLIO-1647  

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [ ] Are there enough tests
- [x] Documentation has been included (for new feature)

## Doc
Some doc is present on plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditIncident.tsx

## Changes

- added title and comment to incident report model + migrations
- added those field in form
- added Other option in types, not displayed in the frontend for now as POLIO-1650 ticket is hold due to the complexity of the feature
- review number inputs logic on the frontend to make it clear and match to this:
<img width="845" alt="Screenshot 2024-08-22 at 12 23 26" src="https://github.com/user-attachments/assets/664d6ca7-02e0-4a48-bd65-7c6a92a4a82c">

## How to test

Open a stock variation for a country with existing data. 
<img width="350" alt="Screenshot 2024-08-22 at 16 03 04" src="https://github.com/user-attachments/assets/fb379ed0-2a2a-4757-aa9e-cfd638030885">

open the details for one entry and click on stock variation.
Select Incident reports tab .
Edit and create reports, check it matches the sheet
Add title and comment and save.

## Print screen / video

<img width="809" alt="Screenshot 2024-08-22 at 15 58 01" src="https://github.com/user-attachments/assets/6e6346f5-af69-47df-8bee-da22ef7fbb66">
<img width="977" alt="Screenshot 2024-08-22 at 15 57 55" src="https://github.com/user-attachments/assets/b1643050-134a-415c-be56-36b5a7cc9581">
<img width="742" alt="Screenshot 2024-08-22 at 15 57 49" src="https://github.com/user-attachments/assets/16b2769e-52f8-40e2-833a-419735c53978">
<img width="1646" alt="Screenshot 2024-08-22 at 15 57 43" src="https://github.com/user-attachments/assets/e7a21147-5d44-45ec-ae49-81f81aac36ef">

beygorghor committed:
POLIO-1645 POLIO-1646 POLIO-1650

- Adding Other incident report type
- adding required field Title
- adding non-required field comment

beygorghor committed:
display OTHER incident reports in stocks

beygorghor committed:
IA-1647-movement-types

- remove other option
- review frontend modal for incident report

beygorghor committed:
black

beygorghor committed:
translations fix

beygorghor committed:
small fixes

quang-le commented on plugins/polio/api/vaccines/stock_management.py:
```diff
@@ -174,6 +175,7 @@ def get_list_of_usable_vials(self):
                 or report.stock_correction == IncidentReport.StockCorrectionChoices.VVM_REACHED_DISCARD_POINT
                 or report.stock_correction == IncidentReport.StockCorrectionChoices.UNREADABLE_LABEL
                 or report.stock_correction == IncidentReport.StockCorrectionChoices.BROKEN
+                or report.stock_correction == IncidentReport.StockCorrectionChoices.OTHER
```
This is another PR no?

quang-le commented on plugins/polio/api/vaccines/stock_management.py:
```diff
@@ -261,6 +263,7 @@ def get_list_of_unusable_vials(self):
                 or report.stock_correction == IncidentReport.StockCorrectionChoices.VACCINE_EXPIRED
                 or report.stock_correction == IncidentReport.StockCorrectionChoices.VVM_REACHED_DISCARD_POINT
                 or report.stock_correction == IncidentReport.StockCorrectionChoices.UNREADABLE_LABEL
+                or report.stock_correction == IncidentReport.StockCorrectionChoices.OTHER
```
same same

quang-le commented on plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditIncident.tsx:
```diff
@@ -38,18 +115,82 @@ export const CreateEditIncident: FunctionComponent<Props> = ({
+        if (!incident) return 0;
+        const movementType = incidentReportConfig[incident.stock_correction];
+        return movementType === 'inventory' ? 0 : incident.usable_vials;
+    }, [incident]);
```
Couldn't you use `useMemo` instead, to pass the return value directly to formik?
Honest question, I can imagine that we would want the value passed to `initialValues` to be static, and not "stateful".

Alternately, I think you can move the function outside of the component:
```typescript
const getInitialMovement = (incident) => {
        if (!incident) return 0;
        const movementType = incidentReportConfig[incident.stock_correction];
        return movementType === 'inventory' ? 0 : incident.usable_vials;
    }
```

quang-le commented on plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditIncident.tsx:
```diff
@@ -59,6 +200,31 @@ export const CreateEditIncident: FunctionComponent<Props> = ({
+            return formatMessage(MESSAGES.missingMovement, { movement });
+        }
+        return '';
+    };
```
Can be moved out of the component

quang-le commented on plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditIncident.tsx:
```diff
@@ -59,6 +200,31 @@ export const CreateEditIncident: FunctionComponent<Props> = ({
+            default:
+                return MESSAGES.movement;
+        }
+    };
```
same same

quang-le commented on plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditIncident.tsx:
```diff
@@ -100,22 +277,77 @@ export const CreateEditIncident: FunctionComponent<Props> = ({
+                        </Typography>
+                    </Box>
+                )}
+                {currentMovementType && currentMovementType === 'inventory' && (
```
`currentMovementType` cannot be equal to `'inventory'` if it's falsy 😉

quang-le commented on plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Table/columns.tsx:
```diff
@@ -220,6 +220,14 @@ export const useIncidentTableColumns = (
+                sortable: true,
+                Cell: settings =>
+                    settings.row.original.title || textPlaceholder,
+            },
```
You could use `BreakWordCell` here

quang-le reviewed (changes requested):
Some movements are not translated in the usable/unusable tables:
![Screenshot 2024-08-23 at 11 56 03](https://github.com/user-attachments/assets/5d0cd073-a154-4d30-a9ad-12101416b2d5)

Otherwise nothing really blocking front-end side

Changes should be made in the backend for the movements that remove usable vials without adding the equivalent unusable vials.
I don't know if it's ok to merge the front-end changes without the backend fix, but without it, such movements are not shown in the main tables :

https://github.com/user-attachments/assets/6acbb5d6-1541-4c0d-8855-48922f04cd38

Jira ticket created for this issue: https://bluesquare.atlassian.net/browse/POLIO-1655

Adjacent to this PR, the table of incident reports is not very legible, since the columns "usable vials" and unusable vials show In and OUT movements in the same column:

![Screenshot 2024-08-23 at 10 21 07](https://github.com/user-attachments/assets/c52f54d9-dc25-4a59-9d42-48410f572795)

quang-le committed:
POLIO-1655: replace check on unusable vials where needed

- LOSSES, STEALING, BROKEN and RETURN incident types now check only usable vials and don't update unusable vials
- Commented out OTHER, since it's still being specced

quang-le committed:
POLIO-1655: fix failing test on vaccine authorizations

beygorghor committed:
pr review

beygorghor committed:
adding translation on stock details

quang-le committed:
POLIO-1655: fix vaccine stock calculations
- vials received: only take into account vials from arrival reports. Exclude those found in physicla inventory
- vials used: aggregate vials from form A outgoing movement i.o using the difference between received and usable vials
- Add methods to VaccineStockCalculator to make these calculations
- Refactor existing methods (mostly because the new methods just extract some code from the old ones)
- Update and comment tests

quang-le committed:
Merge branch 'POLIO-1636-incident-reports' into POLIO-1655_adapt_stock-backend

quang-le committed:
POLIO-1636: black formatting

quang-le commented on plugins/polio/js/src/domains/VaccineModule/StockManagement/hooks/useGetMovementDescription.ts:
```diff
@@ -0,0 +1,17 @@
+export const useGetMovementDescription = () => {
+    const { formatMessage } = useSafeIntl();
+
+    return (movementType: IncidentReportFieldType, movement: number) => {
+        if (movementType === 'plainMovement') {
+            return formatMessage(MESSAGES.plainMovement, { movement });
+        }
+        if (movementType === 'missingMovement') {
+            return formatMessage(MESSAGES.missingMovement, { movement });
+        }
+        return '';
+    };
+};
```
That should be `useCallback` no?

beygorghor replied:
This is o, fact returning a string, so not necessary i would say

quang-le committed:
POLIO-1655: black formatting

quang-le committed:
Update plugins/polio/api/vaccines/stock_management.py

Co-authored-by: Marc Hertzog <marc@marcarea.com>

quang-le committed:
Update plugins/polio/api/vaccines/stock_management.py

Co-authored-by: Marc Hertzog <marc@marcarea.com>

quang-le committed:
POLIO-1655: use settings.DNS_DOMAIN for vaccine authorizations emails

quang-le committed:
Merge branch 'POLIO-1655_adapt_stock-backend' of https://github.com/BLSQ/iaso into POLIO-1655_adapt_stock-backend

quang-le committed:
Merge pull request #1576 from BLSQ/POLIO-1655_adapt_stock-backend

POLIO-1655 adapt stock backend

quang-le committed:
POLIO-1636: cleanup prints

quang-le committed:
POLIO-2636: cleanup print
